todo_app/user.py

```diff
@@ -30,7 +41,7 @@ def role(self):
         return self._role if self._role is not None else Roles.reader
 
     @staticmethod
-    def check_permission(level):
+    def check_permission(level, action: Actions, app: Flask = None):
```

Comment:
We can access the `app` variable with `current_app` imported from Flask, which should correctly find the app without needing to pass it about directly. Either of these approaches do drive us towards having lots of functions which are inherently dependent on Flask though, so it might be that we provide an internal function for `get_logger` or similar that wraps that at least

Reply:
How might a get_logger function work? I did consider a wrapper around the logger that automatically passes things like user ID by default. I thought it would add extra complexity due to the different logs passing different values depending on the context.

Reply:
Ah this is the question - many different options!

It might be that we just want a function that returns a logger, which uses the app logger if available, or falls back to something else for e.g. unit tests where it isn't:
```python
def get_custom_logger():
    if current_app:
        return current_app.logger
    else:
        return getLogger()
```

But if we want it to auto-populate some fields, we could take that further by creating a wrapper class, and adding our own `info` methods etc. that in turn call the underlying logger/info method, adding in their context

In thinking about this, I did realise I may be re-building something that already exists, and it looks like maybe [LoggerAdapters](https://docs.python.org/2/howto/logging-cookbook.html#adding-contextual-information-to-your-logging-output) are the textbook answer and basically help us achieve that without re-writing the info/warn functions